### PR TITLE
Experiment: Make image_transport_plugins_tests work with clang

### PR DESCRIPTION
**Pull request by lalten**
`❯ bazel test //ros2/test/image_common:image_transport_plugins_tests`

`C++ exception with description "MultiLibraryClassLoader: Could not create object of class type image_transport::RawSubscriber as no factory exists for it. Make sure that the library exists and was explicitly loaded through MultiLibraryClassLoader::loadLibrary()" thrown in the test body.`

<details><summary>Full log, failing test</summary>

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //ros2/test/image_common:image_transport_plugins_tests
-----------------------------------------------------------------------------
Running main() from gmock_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from TestImageTransportPlugins
[ RUN      ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsurePublisherPluginCanBeLoaded
unknown file: Failure
C++ exception with description "MultiLibraryClassLoader: Could not create object of class type image_transport::RawPublisher as no factory exists for it. Make sure that the library exists and was explicitly loaded through MultiLibraryClassLoader::loadLibrary()" thrown in the test body.
[  FAILED  ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsurePublisherPluginCanBeLoaded (10 ms)
[ RUN      ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsureSubscriberPluginCanBeLoaded
unknown file: Failure
C++ exception with description "MultiLibraryClassLoader: Could not create object of class type image_transport::RawSubscriber as no factory exists for it. Make sure that the library exists and was explicitly loaded through MultiLibraryClassLoader::loadLibrary()" thrown in the test body.
[  FAILED  ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsureSubscriberPluginCanBeLoaded (10 ms)
[----------] 2 tests from TestImageTransportPlugins (20 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (20 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsurePublisherPluginCanBeLoaded
[  FAILED  ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsureSubscriberPluginCanBeLoaded

 2 FAILED TESTS
```

</details>

<details><summary>Full log, fixed test</summary>

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //ros2/test/image_common:image_transport_plugins_tests
-----------------------------------------------------------------------------
Running main() from gmock_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from TestImageTransportPlugins
[ RUN      ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsurePublisherPluginCanBeLoaded
[       OK ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsurePublisherPluginCanBeLoaded (11 ms)
[ RUN      ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsureSubscriberPluginCanBeLoaded
[       OK ] TestImageTransportPlugins.WhenRawPluginsAvailable_EnsureSubscriberPluginCanBeLoaded (0 ms)
[----------] 2 tests from TestImageTransportPlugins (11 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (11 ms total)
[  PASSED  ] 2 tests.
```

</details>

### CODE REVIEW

File: examples/.bazelrc [hunk 1]
```diff
@@ -24,3 +24,7 @@ test --build_tests_only
 # as the user configuration should be able to overwrite flags from this file.
 # See https://docs.bazel.build/versions/master/best-practices.html#bazelrc
 try-import %workspace%/user.bazelrc
+
+build --repo_env=CC=clang
+build --repo_env=CXX=clang++
+build --linkopt=-B/usr/bin/ld.lld
```
Comment (mvukov) [inline examples/.bazelrc:30]: Please keep the .bazelrc files identical (CI fails because of this). Then we can see whether gcc build still works.

Comment (lalten) [reply examples/.bazelrc:30]: ah yep, this should be
```suggestion
build --linkopt=-fuse-ld=/usr/bin/ld.lld
```

Comment (lalten) [reply examples/.bazelrc]: The image's clang is installed in https://github.com/actions/runner-images/blob/main/images/linux/scripts/installers/clang.sh#L14,
so it does jsut install clang via apt, which means we have https://ubuntu.pkgs.org/22.04/ubuntu-universe-amd64/lld-14_14.0.0-1ubuntu1_amd64.deb.html
and the path `/usr/lib/llvm-14/bin/lld` may work.

Comment (lalten) [reply examples/.bazelrc]: OK CI is still on Ubuntu20.04, so I set it to the default lld-11 path.

Comment (lalten) [reply examples/.bazelrc]:
```
lld is a generic driver.
Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
I guess that's not how that works then :sweat_smile:

Comment (lalten) [reply examples/.bazelrc]: Leaving it away does the trick. Interestingly, this does not work on my laptop, but I'm not running Ubuntu.

**Comment by lalten** [comment]
`//ros2/test/image_common:image_transport_plugins_tests                   PASSED in 0.1s`
Some other tests still fail with clang, but the image_transport plugin workaround is working.

**Comment by mvukov** [comment]
Thanks, we can close this PR now and some doc will be in #78.

**Pull request closed without merging**